mlflow/tensorflow/__init__.py

```diff
@@ -769,6 +787,39 @@ def train(original, self, *args, **kwargs):
 
         result = original(self, *args, **kwargs)
 
+        def _extract_data_from_input_fn():
+            def _extract_dict_or_tensor(next_input):
+                input_ex_dict_tensor = None
+                if isinstance(next_input, dict):
+                    input_ex_dict_tensor = {
+                        k: _extract_tensor_as_numpy_input_example_slice(v)
+                        for k, v in next_input.items()
+                    }
+                elif isinstance(next_input, tensorflow.Tensor):
+                    input_ex_dict_tensor = _extract_tensor_as_numpy_input_example_slice(next_input)
+                return input_ex_dict_tensor
+
+            input_training_data = kwargs.get("input_fn")()
+            input_features = None
+            if isinstance(input_training_data, tuple):
+                input_features = _extract_dict_or_tensor(input_training_data[0])
+            elif isinstance(input_training_data, tensorflow.data.Dataset):
+                train_iter = iter(input_training_data)
+                next_input = next(train_iter)[0]
```

Comment:
For my own understanding, why do we take the first element of the resulting output and discard the others?

From the docs for https://www.tensorflow.org/api_docs/python/tf/data/Dataset, I see the following:

```
Element: A single output from calling next() on a dataset iterator. Elements may be nested structures containing multiple components. For example, the element (1, (3, "apple")) has one tuple nested in another tuple. The components are 1, 3, and "apple".
```

This seems to imply that `next()` may yield tuple or another non-scalar result. By only extracting the first element, are we missing some of the dataset's structure?

Reply:
As seen [here](https://www.tensorflow.org/api_docs/python/tf/estimator/Estimator#train): "A [tf.data.Dataset](https://www.tensorflow.org/api_docs/python/tf/data/Dataset) object: Outputs of Dataset object must be a tuple (features, labels) with same constraints as below."

The first item of that tuple is 'x' entry which we need to extract as an input example.

Reply:
Thanks! Can we leave an inline comment explaining this?